### PR TITLE
[Flashscript] Version 2.7.0

### LINKED ISSUE #589: regular expressions strings
Only a minor aesthetic, yet perhaps can be more

    hostcheck = re.search(r"\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}", host)

shouldn't regexp string be prefixed by **r** so the backslashes are preserved?

Another:

     logger.trace(f"\n…

inserts a newline before the string, doesn't it?

### PR DESCRIPTION
**Pull request by andyblac**
- move run block into Main class, move config, device_scan, manual_hosts into own functions, better handling of webserver shutdown.
- add support for RGBW2 (both stock color modes).
- add support for ShellyMotion (list info only, must use manual IP's and use --timeout 240, as motion needs longer to wake and be detected)
- add -dn | --device-name filter to limit scan to include term.
- add '-f | --flash', '--app-version', '--timeout'
- add battery info for those devices that support it.
- python formatting and cleanup, more optimisations
- add regex literal to re.search lines, suggestion by juanfal thanks. fixes #589
- fix keyboard interrupt for manual hosts.
- if device on stock, but not on latest version, nolonger show latest shock version info, but HomeKit version available.

### CODE REVIEW

**Comment by andyblac** [comment]
don't merge yet, i'll mark for review once ready.

File: tools/flash-shelly.py [hunk 1]
```diff
@@ -269,6 +269,18 @@ def get_device_info(self, is_flashing=False):
...
+    if self.fw_type == 'homekit':
+      return True
+    else:
+      return False
+
+  def is_stock(self):
+    if self.fw_type == 'stock':
+      return True
+    else:
+      return False
```
Comment (timoschilling) [inline tools/flash-shelly.py:282]:
```suggestion
  def is_homekit(self):
    return self.fw_type == 'homekit'

  def is_stock(self):
    return self.fw_type == 'stock'
```

File: tools/flash-shelly.py [hunk 2]
```diff
@@ -290,19 +302,19 @@ def set_force_version(self, version):
   def get_current_version(self, is_flashing=False): # used when flashing between formware versions.
     if not self.get_device_info(is_flashing):
       return None
-    if self.fw_type == 'homekit':
+    if self.is_homekit:
```
Comment (timoschilling) [inline tools/flash-shelly.py:305]:
```suggestion
    if self.is_homekit():
```

File: tools/flash-shelly.py (hunk 1 above)
Comment (andyblac) [reply tools/flash-shelly.py:282]: thx

File: tools/flash-shelly.py (hunk 2 above)
Comment (andyblac) [reply tools/flash-shelly.py:305]: thx

**Comment by rojer** [comment]
i think saving and deleting config should not be handled by the script. if user wants to provide a config, they should write one. same, if a user wants to get rid of the config, there's `rm` or `del`.
i'm still not convinced that config is even needed, for average user.
i'm sure advanced users are capable of storing different invocations of the script.
@timoschilling you requested it, can you elaborate?

**Comment by andyblac** [comment]
> i think saving and deleting config should not be handled by the script. if user wants to provide a config, they should write one. same, if a user wants to get rid of the config, there's `rm` or `del`.
> i'm still not convinced that config is even needed, for average user.
> i'm sure advanced users are capable of storing different invocations of the script.
> @timoschilling you requested it, can you elaborate?

cant be done ALL flag have to be included in config, not option you want, as it use configparser it sets new default from the config. i can remove delete option np though.

**Comment by andyblac** [comment]
sample of a config section
```
[motionsensors]
mode = keep
info_level = 3
fw_type_filter = all
model_type_filter = all
device_name_filter = all
do_all = False
quiet_run = False
list = True
exclude = 
dry_run = False
silent_run = False
version = 
variant = 
local_file = 
hap_setup_code = 
network_type = 
ipv4_ip = 
ipv4_gw = 
ipv4_mask = 
ipv4_dns = 
verbose = 3
timeout = 360
log_filename = 
reboot = False
hosts = 192.168.3.20 192.168.3.43 192.168.3.46 192.168.3.47
```

**Comment by rojer** [comment]
> cant be done, as ALL flags have to be included in config, not option you want

and yet it has to be. otherwise every new option will break existing user configs.
if configparser can't do defaults (i find it hard to believe), don't use configparser, use something else.
YAML is a nice, user friendly config language but unfortunately an extra dependency.
JSON is in the standard library so it's tempting to use but it's a terrible choice for configs, let's not.

**Comment by rojer** [comment]
but back to the question: why do we need this at all?

**Comment by andyblac** [comment]
this is code

```
    parser.set_defaults(mode=config.get(args.config, 'mode'), info_level=config.getint(args.config, 'info_level'), fw_type_filter=config.get(args.config, 'fw_type_filter'),
                        model_type_filter=config.get(args.config, 'model_type_filter'), device_name_filter=config.get(args.config, 'device_name_filter'),
                        do_all=config.getboolean(args.config, 'do_all'), quiet_run=config.getboolean(args.config, 'quiet_run'), list=config.getboolean(args.config, 'list'),
                        exclude=config.get(args.config, 'exclude'), dry_run=config.getboolean(args.config, 'dry_run'), silent_run=config.getboolean(args.config, 'silent_run'),
                        version=config.get(args.config, 'version'), variant=config.get(args.config, 'variant'), local_file=config.get(args.config, 'local_file'),
                        hap_setup_code=config.get(args.config, 'hap_setup_code'), network_type=config.get(args.config, 'network_type'), ipv4_ip=config.get(args.config, 'ipv4_ip'),
                        ipv4_gw=config.get(args.config, 'ipv4_gw'), ipv4_mask=config.get(args.config, 'ipv4_mask'), ipv4_dns=config.get(args.config, 'ipv4_dns'),
                        verbose=config.getint(args.config, 'verbose'), timeout=config.get(args.config, 'timeout'), log_filename=config.get(args.config, 'log_filename'),
                        reboot=config.getboolean(args.config, 'reboot'), hosts=config.get(args.config, 'hosts').split())
```

**Comment by andyblac** [comment]
> but back to the question: why do we need this at all?

for example, he shelly motions are not found by the scanner bue to the way they are to preserve battery, the only way to wake the devcie for web access or rpc is to connect to it via port 80 on the ip, hostname also do not work, so when using the script with --config {profile name} you can load setting required to handle them in one go, without having to remember their IP's
```
andrews-mbp-wifi:tools andrewblackburn$ ./flash-shelly.py --timeout 360 -l -i3 -m=keep 192.168.3.46 192.168.3.47 192.168.3.20 192.168.3.43 --save-config motionsensors
Saving configuration section motionsensors to .flashrc
andrews-mbp-wifi:tools andrewblackburn$ ./flash-shelly.py --config motionsensors
Reading configuration section motionsensors from .flashrc
Looking for Shelly devices...

Host: http://192.168.3.46
Device Name: Hall Motion Sensor
Model: ShellyMotion
Device ID: shellymotionsensor-60A4239A69B0
SSID: ABFi-IoT-2.4
IP: 192.168.3.46
RSSI: -42
Up Time: 1 day, 12:21:37
Battery: 91%
Firmware: Stock 1.1.0 ✔

Host: http://192.168.3.47
Device Name: Landing Motion Sensor
Model: ShellyMotion
Device ID: shellymotionsensor-60A4239368A2
SSID: ABFi-IoT-2.4
IP: 192.168.3.47
RSSI: -33
Up Time: 12 days, 4:14:50
Battery: 94%
Firmware: Stock 1.1.0 ✔

Host: http://192.168.3.20
Device Name: Back Garden Motion Sensor
Model: ShellyMotion
Device ID: shellymotionsensor-60A423976642
SSID: ABFi-IoT-2.4
IP: 192.168.3.20
RSSI: -41
Up Time: 10:55:41
Battery: 95%
Firmware: Stock 1.1.0 ✔

Host: http://192.168.3.43
Device Name: Front Garden Motion Sensor
Model: ShellyMotion
Device ID: shellymotionsensor-60A4239752B4
SSID: ABFi-IoT-2.4
IP: 192.168.3.43
RSSI: -55
Up Time: 14:43:22
Battery: 96%
Firmware: Stock 1.1.0 ✔

Devices found: 4 Upgradeable: 0
```

**Comment by rojer** [comment]
how is this different from having `motionsensors.sh` with `./flash-shelly.py --timeout 360 -l -i3 -m=keep 192.168.3.46 192.168.3.47 192.168.3.20 192.168.3.43` ?

**Comment by andyblac** [comment]
> how is this different from having `motionsensors.sh` with `./flash-shelly.py --timeout 360 -l -i3 -m=keep 192.168.3.46 192.168.3.47 192.168.3.20 192.168.3.43` ?

you can't overrule those bu extra command options
```motionsensors.sh -i2`

**Comment by rojer** [comment]
sure i can:

`./flash-shelly.py --timeout 360 -l -i3 -m=keep "$@" 192.168.3.46 192.168.3.47 192.168.3.20 192.168.3.43`

**Comment by andyblac** [comment]
> > cant be done, as ALL flags have to be included in config, not option you want
> 
> and yet it has to be. otherwise every new option will break existing user configs.
> if configparser can't do defaults (i find it hard to believe), don't use configparser, use something else.
> YAML is a nice, user friendly config language but unfortunately an extra dependency.
> JSON is in the standard library so it's tempting to use but it's a terrible choice for configs, let's not.

is it not only down file format, but argparser. you have to set new defaults from loaded config however you save / load it.

**Comment by andyblac** [comment]
> sure i can:
> 
> `./flash-shelly.py --timeout 360 -l -i3 -m=keep "$@" 192.168.3.46 192.168.3.47 192.168.3.20 192.168.3.43`

try it i bet it complains about 2 lots of same option ie `i3 i2`

**Comment by rojer** [comment]
nope, it just uses the last value

**Comment by andyblac** [comment]
tbh i do either way, but personally i dont see an issue if it not default.

**Comment by rojer** [comment]
> i dont see an issue if it not default.

sorry, i can;t parse that :) can you rephrase?

**Comment by andyblac** [comment]
> > i dont see an issue if it not default.
> 
> sorry, i can;t parse that :) can you rephrase?

i do not see use config's being issue, as long as option 2 in voting is not default.

**Comment by rojer** [comment]
the issue is of extra complexity - in my opinion it is not warranted.
features have cost, they need to be documented, maintained, tested and re-tested, etc.
we just don't need this, imo.

**Comment by andyblac** [comment]
> the issue is of extra complexity - in my opinion it is not warranted.
> features have cost, they need to be documented, maintained, tested and re-tested, etc.
> we just don't need this, imo.

i dont see how configs add's extra complexity, it is simply just setting new defaults. ?

**Comment by rojer** [comment]
it's code complexity. a fair bit of code is dedicated to parsing config and getting values from it. every flag has to be duplicated there, you need to remember that when adding flags, failing to do that will break config functionality, silently.
plus, in its current form it's unusable (due to requiring having all the flags set), this will need to change.

so, my vote is against having the config at all, we just don't need it at this time.

**Comment by andyblac** [comment]
> plus, in its current form it's unusable (due to requiring having all the flags set), this will need to change.
> 
> so, my vote is against having the config at all, we just don't need it at this time.

i dont how you IF you want to keep command line prioity.

**Comment by rojer** [comment]
i'm sure there's a way but i don't even want to spend time looking because it's irrelevant.
i think the whole things should go, there's no point in even spending time on it.
i bet at this point more time has been spent between the three of us developing and discussing it than it would have ever saved.

i'll let @timoschilling put forward his argument though, maybe i'm missing something.

**Comment by timoschilling** [comment]
@rojer sure your way works as well, the config way is just more comfortable and a quite common pattern.

**Comment by rojer** [comment]
> @rojer sure your way works as well, the config way is just more comfortable and a quite common pattern.

i know configs are common, just not for this.
so, the final verdict is: no config, for now.

**Comment by andyblac** [comment]
ok i'll remove.

File: tools/flash-shelly.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python3
 """
-Copyright (c) 2021 Andrew Blackburn & Deomid "rojer" Ryabkov
+Copyright (c) 2021 Andrew Blackburn, Deomid "rojer" Ryabkov & Timo Schilling
```
Comment (rojer) [inline tools/flash-shelly.py:3]: "Shelly-HomeKit Contributors"

**Comment by andyblac** [comment]
@rojer @timoschilling please give this a test mainly 

flashing from factory stock to hk
flashing RGBW2 from stock to hk, both white / color modes
reverting from hk to stock.

if any issues found please run again with `--log-file debug.log` flag and post file.

once were happy, i'll start cleaning up verbose log info.

thx

**Comment by andyblac** [comment]
@rojer ready for review.
